acra-censor/handlers/handlers_util.go

```diff
@@ -125,711 +171,1558 @@ func NormalizeAndRedactSQLQuery(sql string) (normalizedQuery string, redactedQue
-	// collect only SelectExpr, From, Where, OrderBy ... nodes without their children
-	queryTopNodes, err := getTopNodes(queryNodes[0])
-	if err != nil {
+func handleSelectStatement(query, pattern sqlparser.Statement) bool {
+	match := false
+	querySelectNode, ok := query.(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	patternTopNodes, err := getTopNodes(patternNodes[0])
-	if err != nil {
+	patternSelectNode, ok := pattern.(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	hasStar := false
-	if selectExpr, ok := patternNodes[0].(*sqlparser.Select); ok {
-		hasStar = starFound(selectExpr.SelectExprs)
-	}
-
-	for i := 0; i < len(queryTopNodes); i++ {
-		patternNode := patternTopNodes[i]
-		queryNode := queryTopNodes[i]
-		switch patternNode.(type) {
-		case sqlparser.SelectExprs:
-			if hasStar {
-				// if select * in pattern then skip columns in query
-				continue
-			}
-		case *sqlparser.Where:
-			if _, ok := queryNode.(*sqlparser.Where); !ok {
-				return false
-			}
-			if !handleWhereNode(patternNode.(*sqlparser.Where), queryNode.(*sqlparser.Where)) {
-				return false
-			}
-			continue
-		}
 
-		if !reflect.DeepEqual(patternNode, queryNode) {
-			return false
-		}
+	// check %%SELECT%%
+	if reflect.DeepEqual(pattern, SelectPatternStatement) {
+		return true
 	}
-	return true
-}
 
-// handleLimitValuePattern evaluates LIMIT=%%VALUE%% and OFFSET=%%VALUE%% patterns
-func handleLimitValuePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
-	querySelect, ok := queryNodes[0].(*sqlparser.Select)
-	if !ok {
+	match = matchSelectCache(querySelectNode.Cache, patternSelectNode.Cache)
+	if !match {
 		return false
 	}
-	patternSelect, ok := patternNodes[0].(*sqlparser.Select)
-	if !ok {
+	match = matchSelectComments(querySelectNode.Comments, patternSelectNode.Comments)
+	if !match {
 		return false
 	}
-	queryTopNodes, err := getTopNodes(queryNodes[0])
-	if err != nil {
+	match = matchSelectDistinct(querySelectNode.Distinct, patternSelectNode.Distinct)
+	if !match {
 		return false
 	}
-	patternTopNodes, err := getTopNodes(patternNodes[0])
-	if err != nil {
+	match = matchSelectHints(querySelectNode.Hints, patternSelectNode.Hints)
+	if !match {
 		return false
 	}
-
-	for i := 0; i < len(queryTopNodes); i++ {
-		patternNode := patternTopNodes[i]
-		queryNode := queryTopNodes[i]
-
-		switch patternNode.(type) {
-		case *sqlparser.Limit:
-			if !matchLimit(querySelect.Limit, patternSelect.Limit) {
-				return false
-			}
-		default:
-			if !reflect.DeepEqual(queryNode, patternNode) {
-				return false
-			}
+	match = matchSelectSelectExprs(querySelectNode.SelectExprs, patternSelectNode.SelectExprs)
+	if !match {
+		return false
+	}
+	match = matchSelectFrom(querySelectNode.From, patternSelectNode.From)
+	if !match {
+		return false
+	}
+	match = matchSelectWhere(querySelectNode.Where, patternSelectNode.Where)
+	if !match {
+		// Check %%WHERE%% pattern
+		if isWherePattern(patternSelectNode.Where) {
```

Comment:
imo better to check most general pattern first instead concrete. it's easier to check simple pattern that avoid other checks of conditions in `where`. after that, it has sense to check more concrete and complex patterns

Reply:
if %%WHERE%% true - we should stop further checks and return true